libraries/abstractions/transport/secure_sockets/transport_secure_sockets.c

```diff
@@ -0,0 +1,511 @@
+
+    if( ( pMessage == NULL ) ||
+        ( bytesToSend <= 0UL ) ||
+        ( pNetworkContext == NULL ) )
+    {
+        LogError( ( "Invalid parameter: pMessage=%p, bytesToSend=%lu, pNetworkContext=%p",
+                    pMessage, bytesToSend, pNetworkContext ) );
+        bytesSent = SOCKETS_EINVAL;
+    }
+    else if( ( pNetworkContext != NULL ) && ( pNetworkContext->pContext == NULL ) )
+    {
+        LogError( ( "Invalid parameter: pNetworkContext->pContext cannot be NULL." ) );
+        bytesSent = SOCKETS_EINVAL;
+    }
+    else
+    {
+        transportSocketStatus = SOCKETS_Send( ( Socket_t ) pNetworkContext->pContext,
+                                              pMessage,
+                                              bytesToSend,
+                                              0 );
+
+        /* If an error occurred, a negative value is returned. @ref SocketsErrors. */
+        if( transportSocketStatus >= 0 )
+        {
+            bytesSent = transportSocketStatus;
+            LogInfo( ( "Successfully sent %ld bytes over network.", transportSocketStatus ) );
+        }
+        else
+        {
+            LogError( ( "Failed to send data over network. BytesRequested=%ld.", transportSocketStatus ) );
+            bytesSent = SOCKETS_SOCKET_ERROR;
+        }
+    }
+
+    return bytesSent;
+}
+
+/*-----------------------------------------------------------*/
+
+int32_t SecureSocketsTransport_Recv( const NetworkContext_t * pNetworkContext,
+                                     void * pBuffer,
+                                     size_t bytesToRecv )
+{
+    int32_t transportSocketStatus = 0;
+    int32_t bytesReceived = SOCKETS_SOCKET_ERROR;
+    uint8_t * pRecvBuffer = ( uint8_t * ) pBuffer;
+
+    if( ( pBuffer == NULL ) ||
+        ( bytesToRecv <= 0UL ) ||
+        ( pNetworkContext == NULL ) )
+    {
+        LogError( ( "Invalid parameter: pBuffer=%p, bytesToRecv=%lu, pNetworkContext=%p",
+                    pBuffer, bytesToRecv, pNetworkContext ) );
+        bytesReceived = SOCKETS_EINVAL;
+    }
+    else if( ( pNetworkContext != NULL ) && ( pNetworkContext->pContext == NULL ) )
+    {
+        LogError( ( "Invalid parameter: pNetworkContext->pContext cannot be NULL." ) );
+        bytesReceived = SOCKETS_EINVAL;
+    }
+    else
+    {
+        transportSocketStatus = SOCKETS_Recv( ( Socket_t ) pNetworkContext->pContext,
+                                              pRecvBuffer,
+                                              bytesToRecv,
+                                              0 );
+
+        if( transportSocketStatus == SOCKETS_EWOULDBLOCK )
+        {
+            /* The return value EWOULDBLOCK means no data was received within
+             * the receive timeout. */
+            bytesReceived = 0;
+        }
+        else if( transportSocketStatus < 0 )
+        {
+            LogError( ( "Failed to receive data over network. SocketsStatus=%d", transportSocketStatus ) );
```

Comment:
Same question about why we don't return the negative status?

Reply:
This API return how much byte it received.
bytesReceived is initialized with "int32_t bytesReceived = SOCKETS_SOCKET_ERROR;"

There are 3 cases will happen. 

Case1:
Timeout happen and gets SOCKETS_EWOULDBLOCK from SOCKETS_Recv.
#define SOCKETS_EWOULDBLOCK              ( -11 )   /*!< A resource is temporarily unavailable. */
Same as [_networReceiveTask in iot_network_freertos.c](https://github.com/aws/amazon-freertos/blob/integration/csdk_v4_beta2/libraries/abstractions/platform/freertos/iot_network_freertos.c#L174) We don't return error.

Case2:
SOCKETS_Recv did return negative value,  bytesReceived is initialized with negative value and it will return.

Case3:
SOCKETS_Recv success and bytesReceived returns the length it received.

Reply:
simplified the logic to avoid confusion in 3e78d0f4ec89db890c2801c1d13e61bceb12f869